### DISCUSSION TITLE
Some UX Changes / Wishlist

### DISCUSSION THREAD

**Discussion by modellking** [arthurmatthew/clicktrack#76] [Ideas]
## 1 Navigation
When playing a set and switching songs one must navigate around the menu a bit to get to the next song. To make this easier a few things should be changed (imo):
1. Fix #75
2.a. Either add Next / Previous buttons to the Metronome Edit page
2.b Or Add a Play/Pause Button to the Overview
2.c Or redesign the Clicktracks and Clicktrack Page to be in one View

## 2 Click sound (Submitted)
In my local church the need for an "atonal" Click was brought up.
1. I am investigating #62 to get that in with a few defaults

## 3 Import options
While preparing a set importing via external formats / links is a really nice feature
1. Either allow a link format to add a Metronome, paired with some injectabe code to be used with e.g. ChurchTools / SongSelect to import Metronomes from their calalog.
2. Or Add File Import supporting more complex formats such as onsong.
3. Also at some point we should fix the renaming UI

## 4 Basic Embeds
During a Set one often uses a Leadsheet or sone kind of pdf / Embeddable File with each song. It'd be nice to consolidate that into one tool at least rudamentarily with pdfs.
1. Add an embed with a pdf upload

## 5 Table for Accents/Sounds
I've worked on the Accents a bit with #74 . When #62 is working instead of those two choices being seperate, other tools have a sequencer to step through the measure (like a matrix/table) where each column allows one to costumize the sound at that click. This looks similar to drum machines: https://www.wavealchemy.co.uk/wp-content/uploads/2022/10/Drumvolution_Screenshot_Sequencer.jpg though the number of columns depends on the kind of metronome (4 for 4/4)

## 6 Support other Section types
Add support for AutoMD and Audio clicktracks.
This could look simply like a section that just plays the linked audiofile

## 7 Technical Improvments
There are a few things That could likely help the long-term maintainability of the app
1. Move Storage to React Query as a Caching-Layer. This streamlines the data handling a lot and should reduce a lot of props-drilling

**Comment by modellking** [comment]
Also please let's talk about a license - While you may assume that I license any code that I provide under MIT, I would like to have assurance that any new code you add remains usable to me.

**Reply by arthurmatthew** [reply]
I'll definitely look into a license. My goal for this project is to eventually host a full-stack application in which users can register an account to have access to their clicktracks anywhere. This is where it gets confusing for me, as I'm not sure which license is fair to use in this situation. I've only developed standalone projects in the past.

**Reply by modellking** [reply]
You want to earn something with this project?
Honestly you'd have a bit of a problem: There are a lot of free metronomes out there. And most of them are mobile anyway and people just take those with them, where they need them
You could maybe work on a App Front end (I think i heard something about PWAs becoming advertisable over the AppStores) and charge 1$ per Install or go ad-supported... Maybe offer an ad-free experience for 1$/month or something... Or setup a patreon?
I think the largest value you'll get out of this project is likly it being a nice portfolio piece.

Independent of that - since your target audience are mostly non-devs and your Service being the delivery and hosting, you could just keep the project FOSS-licensed

**Reply by arthurmatthew** [reply]
Agreed, I've added an MIT license

Honestly, I don't care much about the profits. At most, I'd like to be able to afford to pay for storage of Clicktracks (if this project ever reaches that level of users). It's mainly proof of concept, it is more appealing on a portfolio like you said, if I "created a full-stack web application which generated revenue." 

The ideas you listed are interesting though. I'd like to look more into that once the project is ready for full public release. Right now, development is obviously the number one priority.

**Reply by modellking** [reply]
The storage for the click tracks should be well within any providers free tier - with 2GB for example we could currently store 20 clicktracks for 400,000 users creating 250 setlists each at least! https://uploadthing.com/

**Reply by arthurmatthew** [reply]
Sounds great, my apologies as I've been busy, but the v0.3.0 branch should be fine to work with currently when using the demo Clicktrack. I should be free soon and I can work on implementing any changes with the backend at that point. I'm also in the process of merging the latest main branch changes, which I'll do tonight

**Comment by arthurmatthew** [comment]
I love these ideas, and I'm very glad that I'm receiving feedback & input. Currently, all the most recent development has been done in the `v0.3.0` branch. I'm going to make sure everything is suitable for development with the cloud storage system I am working on, and then I think its best if development continues in that branch. 

Once I finish that I'll look more closely at the ideas you've listed. I'd love to go into more detail with the other section types and the import options. And the embeddable files. Certain ideas, like the drum machine/accent table, I have attempted to work on, however that was around the time that I had to take a break from development. I'll open up some feature requests to note down the ideas that I have once they come back to me. I may have mentioned it vaguely in #45 .

I'll get back to you hopefully soon!
